core/src/main/java/org/apache/brooklyn/core/effector/composite/AbstractCompositeEffector.java

```diff
@@ -0,0 +1,184 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.brooklyn.core.effector.composite;
+
+import java.util.Map;
+import java.util.Set;
+
+import org.apache.brooklyn.api.effector.Effector;
+import org.apache.brooklyn.api.entity.Entity;
+import org.apache.brooklyn.api.entity.EntityLocal;
+import org.apache.brooklyn.core.effector.AddEffector;
+import org.apache.brooklyn.core.effector.EffectorBody;
+import org.apache.brooklyn.util.collections.CollectionFunctionals;
+import org.apache.brooklyn.util.core.config.ConfigBag;
+import org.apache.brooklyn.util.guava.Maybe;
+import org.apache.brooklyn.util.text.Strings;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.annotations.Beta;
+import com.google.common.collect.Iterables;
+
+@Beta
+public abstract class AbstractCompositeEffector extends AddEffector {
+
+    private static final Logger LOG = LoggerFactory.getLogger(AbstractCompositeEffector.class);
+
+    public AbstractCompositeEffector(Effector<?> effector) {
+        super(effector);
+    }
+
+    @Override
+    public void apply(EntityLocal entity) {
+        Maybe<Effector<?>> effectorMaybe = entity.getEntityType().getEffectorByName(effector.getName());
+        if (!effectorMaybe.isAbsentOrNull()) {
+//            Effector<?> original = Effectors.effector(effectorMaybe.get()).name(ORIGINAL_PREFIX + effector.getName()).build();
+//            ((EntityInternal) entity).getMutableEntityType().addEffector(original);
+        }
+        super.apply(entity);
+    }
+
+    protected static abstract class Body<T> extends EffectorBody<T> {
+        protected final Effector<?> effector;
+        protected final ConfigBag params;
+
+        public Body(Effector<?> eff, ConfigBag params) {
+            this.effector = eff;
+            this.params = params;
+        }
+
+        @Override
+        public abstract T call(final ConfigBag params);
+
+        protected String getEffectorName(Object effectorDetails) {
+            String effectorName = null;
+            if (effectorDetails instanceof String) {
+                effectorName = (String) effectorDetails;
+            } else if (effectorDetails instanceof Map) {
+                Map<String,Object> effectorMap = (Map) effectorDetails;
+                Set<String> keys = effectorMap.keySet();
+                if (keys.size() != 1) {
+                    throw new IllegalArgumentException("Effector parameter cannot be parsed: " + effectorDetails);
+                }
+                effectorName = Iterables.getOnlyElement(keys);
+            } else {
```

Comment:
This is the individual effector elements of the list of effectors, they can only be a `String` or a `Map`